ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,360 @@
+ *      disclaimer in the documentation and/or other materials provided
+ *      with the distribution.
+ *
+ *      Neither the name of Sandia nor the names of any contributors may
+ *      be used to endorse or promote products derived from this software
+ *      without specific prior written permission.
+ *
+ *      Neither the name of Open Grid Computing nor the names of any
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ *      Modified source versions must be plainly marked as such, and
+ *      must not be misrepresented as being the original software.
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <limits.h>
+#include <string.h>
+#include <dirent.h>
+#include <coll/rbt.h>
+#include <sys/queue.h>
+#include <time.h>
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "ldms_jobid.h"
+#include "lustre_client.h"
+#include "lustre_client_general.h"
+
+#define _GNU_SOURCE
+
+#define LLITE_PATH "/proc/fs/lustre/llite"
+#define OSD_SEARCH_PATH "/proc/fs/lustre"
+
+ldmsd_msg_log_f log_fn;
+static char producer_name[LDMS_PRODUCER_NAME_MAX];
+static uint64_t component_id;
+static time_t last_refresh;
+static time_t refresh_interval = 30;
+#ifdef ENABLE_JOBID
+bool with_jobid = true; /* for LJI */
+#endif
+
+/* red-black tree root for llites */
+static struct rbt llite_tree;
+
+struct llite_data {
+        char *fs_name;
+        char *name;
+        char *path;
+        char *stats_path;
+        ldms_set_t general_metric_set; /* a pointer */
+        struct rbn llite_tree_node;
+};
+
+static int string_comparator(void *a, const void *b)
+{
+        return strcmp((char *)a, (char *)b);
+}
+
+static struct llite_data *llite_create(const char *llite_name, const char *basedir)
+{
+        struct llite_data *llite;
+        char path_tmp[PATH_MAX]; /* TODO: move large stack allocation to heap */
+        char *state;
+
+        log_fn(LDMSD_LDEBUG, SAMP" llite_create() %s from %s\n",
+               llite_name, basedir);
+        llite = calloc(1, sizeof(*llite));
+        if (llite == NULL)
+                goto out1;
+        llite->name = strdup(llite_name);
```

Comment:
leaks in llite_destroy?

Reply:
leaks no more than the current  lustre sampler for which term is also never called in normal use.